### ISSUE TITLE
Search compressed files within tgz archive

### ISSUE THREAD

**Issue by acelticsfan** [Genivia/ugrep#152]
I have tgz archives that have gzip (.gz) text files within them but I can't find any option to search those as well.  Is this something the tool already does or would that be an enhancement?

I am using the "-zR" (and "-zr") options.

**Comment by genivia-inc** [comment]
Good point. This has been asked before. It is not impossible to implement, but it is rather tricky, because the stack of decoders (decompressors) has to be set up in advance. Having a deeper stack of decoders (each running a thread) than just one will add more overhead. This makes everything run slower. Perhaps an option could be added that presets the number of stacked decoders? Specifying two would decompress a file stored in an already compressed archive. I have to think a bit about this before promising anything.

**Comment by acelticsfan** [comment]
hi, thanks for getting back.  I appreciate the consideration.   Something like that would work for me, and not unnecessarily slow down those who don't need it.

**Comment by genivia-inc** [comment]
I resumed working on this new feature over the past weekend and made some good progress. The approach I have in mind is to launch multiple decompressor threads, each taking the input from the next decompressor. This approach supports searching multi-compressed files and archives, i.e. archives recursively stored within archives or compressed files nested in other compressed files or archives. Also tar nested within tar is handled this way, so there should not be any limitations on the nesting of compressed/archived contents. The only limitation is that the maximum recursive nesting depth must be specified on the command line in advance to make this work, something like `--zdepth=n` for `n` decompressors/extractors such as `--zdepth=2` to decompress/extract twice. This `--zdepth-2` will of course also handle a single compressed file/archive. It simply sets the maximum recursion depth permitted. Not sure about the name of the `--zdepth` option or `--zlevels` or `--zmax` perhaps. Not sure yet what would be most intuitive to use.

The general approach is simple to explain. However, the details to make this work are a bit tricky, because of the multi-threading with mutex, condition variables, and the use of Unix pipes to connect the threads as well as detecting the archive contents to keep decompressing all parts like before, but now in each of the intermediate decompressors.

I have a prototype version running, but this needs improvements. For one, I have to implement the filenames reported in archives, which are of the form "_archive-filename_{_archive-partname_}" to report matches in parts of an archive file like we already have, but this should be extended to recursive forms, such as _archive-filename_{_archive-filename_{_archive-partname_}} to indicate the nesting, or alternatively, _archive-filename_{_archive-filename_}{_archive-partname_}.

**Comment by genivia-inc** [comment]
Testing went well after correcting an issue that caused problems with the query UI hanging while waiting for the decompression threads to terminate when `--zmax` is set larger than 1 (1 is the default, which behaves the same as the current "old" ugrep version). The problem had to do with the thread termination logic required to cancel searches, which happens when the user types in a new pattern in the query UI that requires a new search to replace the current search results.

The query UI is very efficient in that the search and decompression threads do not need to complete their work if only the first part of the results are viewed in the UI. This means that the threads will continue on demand or need to the terminated when the search pattern is changed by the user when entering the search patterns. I want to make sure the query UI is very responsive, even when searching a lot of files and huge files. I made some improvements to make sure this experience is as smooth as possible.

The updated `-z` option and new `--zmax` option:
~~~
    -z, --decompress
            Decompress files to search, when compressed.  Archives (.cpio,
            .pax, .tar and .zip) and compressed archives (e.g. .taz, .tgz,
            .tpz, .tbz, .tbz2, .tb2, .tz2, .tlz, .txz, .tzst) are searched and
            matching pathnames of files in archives are output in braces.  If
            -g, -O, -M, or -t is specified, searches files stored in archives
            whose filenames match globs, match filename extensions, match file
            signature magic bytes, or match file types, respectively.
            Supported compression formats: gzip (.gz), compress (.Z), zip,
            bzip2 (requires suffix .bz, .bz2, .bzip2, .tbz, .tbz2, .tb2, .tz2),
            lzma and xz (requires suffix .lzma, .tlz, .xz, .txz),
            lz4 (requires suffix .lz4),
            zstd (requires suffix .zst, .zstd, .tzst).
    --zmax=NUM
            When used with option -z (--decompress), searches the contents of
            compressed files and archives stored within archives for up to NUM
            recursive levels deep.  By default NUM=1 to only permit searching
            uncompressed files stored in cpio, pax, tar and zip archives;
            compressed files and archives are detected as binary files and are
            effectively ignored.  Specify --zmax=2 to search compressed files
            and archives stored in cpio, pax, tar and zip archives.  NUM may
            range from 1 to 99 for up to 99 decompression and de-archiving
            steps.  Larger NUM values gradually degrade performance.
~~~
Perhaps `--zmax=99` is a bit ridiculous, but it helped with testing the correctness and performance of the ugrep additions. The performance degrades only slightly with increasing `--zmax` values, because each `--zmax` level runs a decompression thread that communicates via pipes with other decompression threads and the ugrep main search thread.

**Comment by genivia-inc** [comment]
Option `--zmax` is included in ugrep v3.5.0.

**Comment by acelticsfan** [comment]
Works great!!!  Thank you
